src/Stoa.ts

```diff
@@ -100,17 +100,26 @@ class Stoa extends WebService
         this.app.post("/preimage_received", this.putPreImage.bind(this));
 
         // Start the server once we can establish a connection to Agora
-        return this.agora.getBlockHeight()
-            .then(
-                (res) => {
-                    logger.info(`Connected to Agora, block height is ${res.toString()}`);
-                    return super.start();
-                },
-                (err) => {
-                    logger.error(`Error: Could not connect to Agora node: ${err.toString()}`);
-                    process.exit(1);
-                }
-            );
+        return new Promise<void>((resolve, reject) => {
+            this.agora.getBlockHeight()
+                .then(
+                    (height) => {
+                        logger.info(`Connected to Agora, block height is ${height.toString()}`);
+                        super.start()
+                            .then(() => {
+                                this.pending = this.pending.then(() => { return this.catchup(height); });
+                                resolve();
+                            })
+                            .catch((err: any) => {
+                                reject(err);
+                            })
+                    },
+                    (err) => {
+                        logger.error(`Error: Could not connect to Agora node: ${err.toString()}`);
+                        process.exit(1);
+                    }
+                );
+        });
```

Comment:
You do not need to wrap this in a `Promise`. It's a redundant pattern which I eliminated in many places in the code base, and we should be careful not to re-introduce it.
```suggestion
        return this.agora.getBlockHeight()
                .then(
                    (height) => {
                        logger.info(`Connected to Agora, block height is ${height.toString()}`);
                        return super.start();
                    })
                    .then(() => {
                        this.pending = this.pending.then(() => { return this.catchup(height); });
                    }),
                    (err) => {
                        logger.error(`Error: Could not connect to Agora node: ${err.toString()}`);
                        process.exit(1);
                    }
                );
        });
```

Promises can be chained. `a.then((result) { return result.action(); })` will chain the `Promise` returned by `action` to the `Promise` returned by `a`. This is much easier to write and read than nesting things, and requires a single catch.
Not that `throw`ing inside a promise is equivalent to `reject`ing it.

Reply:
Note ^ The above code wasn't tested, but the idea is simple: chain, do not nest.

Reply:
Well, I think it's good to use 'then' duplicated.
I accepted the opinion, fixed the code, and made some modifications through the test.